File: sortable_column.py
from typing import Callable, Optional
from nicegui import ui


class SortableColumn(ui.element, component='sortable_column.js'):

    sortable_list = {}

    def __init__(self, *, on_change: Optional[Callable] = None, group:str = None) -> None:
        super().__init__()
        self.on('item-drop', self.drop)
        self.on_change = on_change

        self._classes.append('nicegui-column')
        self._props['group'] = group
        SortableColumn.sortable_list[self.id] = self

    def drop(self, e) -> None:
        if self.on_change:
            self.on_change(e.args['new_index'],e.args['old_index'],SortableColumn.sortable_list[e.args['new_list']],SortableColumn.sortable_list[e.args['old_list']])
        else:
            print(e.args)

    def makeSortable(self) -> None:
        self.run_method('makeSortable')

File: sortable_column.js
import 'https://cdnjs.cloudflare.com/ajax/libs/Sortable/1.15.0/Sortable.min.js'

export default {
  template: `
    <div>
      <slot></slot>
    </div>
  `,
  props: {
    group: String,
  },
  mounted() {
    this.makesortable();
  },
  methods: {
    makesortable() {
      if (this.group === 'None') {
      this.group = this.$el.id;
      }
      Sortable.create(this.$el, {
      group: this.group,
      animation: 150,
      ghostClass: 'opacity-50',
      onEnd: (evt) => this.$emit("item-drop", {
        parent: parseInt(this.$el.id.slice(1)),
        id: parseInt(evt.item.id.slice(1)),
        new_index: evt.newIndex,
        new_list: evt.to,
        old_list: evt.from,
        event: evt,
      }),
      });
    },
  },
};
